# TF-IDF

We'll rely on term frequency times inverted document frequency to measure meaningful similarity between documents. Let's start by generating a matrix for the separate constituent parts of _Stjórn_.

In [41]:
import os,glob,json
import pandas as pd
import numpy as np
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

In [14]:
titles = ['prologue', 'introduction', 'gn', 'ex', 'lv', 'nm', 'dt', 'ios', 'idc', 'rt', '1sm', '2sm', '3rg', '4rg']
tokens = []
for title in titles:
    with open(f"nlp/{title}.txt") as raw:
        document = raw.read().replace('\n', ' ')
        tokens.extend(document.split())

work_indices = {
    'stjorn1': (650,124417),
    'stjorn2': (124417,147678),
    'stjorn3': (147678,156943,160719),
    'stjorn4': (156943,160719)
}

stjorn = dict()
for _work, _range in work_indices.items():
    if len(_range) == 2:
        stjorn[_work] = ' '.join(tokens[_range[0]:_range[1]])
    else:
        stjorn[_work] = ' '.join(tokens[_range[0]:_range[1]] + tokens[_range[2]:])

menota = dict()
for text in glob.glob('../menota/dipl/*txt'):
    ref = os.path.basename(text).replace('.txt', '')
    with open(text) as doc:
        menota[ref] = doc.read().replace('\n', '')

In [55]:
vectorizer = TfidfVectorizer(min_df=1)
model = vectorizer.fit_transform(stjorn.values())
df = pd.DataFrame(cosine_similarity(model), stjorn.keys(), stjorn.keys())
df

stjorn1   stjorn2   stjorn3   stjorn4
stjorn1  1.000000  0.811713  0.508459  0.842683
stjorn2  0.811713  1.000000  0.394915  0.833537
stjorn3  0.508459  0.394915  1.000000  0.469750
stjorn4  0.842683  0.833537  0.469750  1.000000

This tells us _Stjórn III_ is the most distinct of the three. Remarkably, compared to _Stjórn IV_, which covers some of the same ground, it is the least similar constituent text. Perhaps further analysis can tell us how.

First let's add _Konungs skuggsjá_ from Menota, as well as Unger's own edition of the _Norwegian Homily Book_. Fingers crossed that we have got the normalization standard of the former to approach Unger's methods reasonably well.

In [56]:
nhb = ''
for text in glob.glob('../nhb/nlp/*txt'):
    with open(text) as doc:
        nhb = nhb + doc.read().replace('\n', '')
stjorn_plus = []
for v in stjorn.values():
    stjorn_plus.append(v)
stjorn_plus.extend([menota['nks235g_konungs_skuggsja'], nhb])
model = vectorizer.fit_transform(stjorn_plus)
df = pd.DataFrame(cosine_similarity(model), list(stjorn.keys()) + ['ks', 'nhb'], list(stjorn.keys()) + ['ks', 'nhb'])
df

stjorn1   stjorn2   stjorn3   stjorn4        ks       nhb
stjorn1  1.000000  0.786536  0.523933  0.812035  0.242526  0.387017
stjorn2  0.786536  1.000000  0.402757  0.798033  0.228456  0.330674
stjorn3  0.523933  0.402757  1.000000  0.474972  0.696958  0.828841
stjorn4  0.812035  0.798033  0.474972  1.000000  0.244240  0.385240
ks       0.242526  0.228456  0.696958  0.244240  1.000000  0.711172
nhb      0.387017  0.330674  0.828841  0.385240  0.711172  1.000000

Next, let's model all of Menota along with Stjórn. Perhaps we'll leave Unger's _Homily Book_ in alongside the Menota edition, just for comparison's sake.

In [57]:
corpus = []
titles = []
for k,v in stjorn.items():
    titles.append(k)
    corpus.append(v)
titles.append('nhb')
corpus.append(nhb)
for k,v in menota.items():
    titles.append(k)
    corpus.append(v)
model = vectorizer.fit_transform(corpus)
df = pd.DataFrame(cosine_similarity(model), titles, titles)
df

stjorn1   stjorn2   stjorn3  \
stjorn1                                    1.000000  0.735480  0.438539   
stjorn2                                    0.735480  1.000000  0.336237   
stjorn3                                    0.438539  0.336237  1.000000   
stjorn4                                    0.707385  0.679995  0.377355   
nhb                                        0.294787  0.240394  0.784795   
am132_egils_saga                           0.708617  0.655126  0.422176   
am162btheta_njals_saga                     0.272811  0.214320  0.725551   
holmPerg30_langslog                        0.446772  0.455732  0.250823   
am1056IX_konungs_skuggsja_fragment         0.436763  0.394065  0.225803   
am78_kristinrettir                         0.176707  0.176005  0.564880   
am63_heimskringla3                         0.258047  0.213238  0.715046   
dg4-7_strengleikar                         0.254231  0.197455  0.814551   
am132_droplaugasona_saga                   0.650093  0.616557  0.335241   
am132_kormaks_saga                         0.550610  0.519199  0.294156   
nraNorrFragm53_haralds_saga_hardrada       0.449819  0.412809  0.268834   
am132_finnboga_saga                        0.646263  0.583163  0.354532   
am132_fostbraedra_saga                     0.682834  0.631780  0.371648   
lbsFragm82_olafs_saga_helga                0.215891  0.180212  0.553279   
am619_norwegian_homily_book                0.641680  0.601200  0.403148   
am56_landslog                              0.208290  0.201402  0.570985   
holmPerg17_thomass_saga                    0.564643  0.425434  0.571984   
am383I_thorlaks_saga                       0.206670  0.164035  0.441820   
holmPerg4_thidreks_saga                    0.274807  0.249519  0.814304   
am132_njals_saga                           0.729257  0.681656  0.412977   
am36_heimskringla2                         0.291054  0.242034  0.738967   
am544_voluspa                              0.374771  0.349948  0.184756   
am162bkappa_njals_saga                     0.256072  0.234858  0.626575   
am305_landslog                             0.192328  0.184941  0.551796   
am132_olkofra_thattr                       0.558903  0.537069  0.390822   
nraNorrFragm54_sverris_saga                0.343951  0.317555  0.778280   
nraNorrFragm55B_hakonar_saga               0.186096  0.183713  0.536831   
gks2365_voluspa                            0.117927  0.106593  0.398113   
am243balpha_konungs_skuggsja               0.215042  0.195268  0.600988   
nraNorrFragm51_fagrskinna                  0.116682  0.114490  0.489704   
am132_viga-glums_saga                      0.711795  0.662292  0.399701   
am279a_gragas                              0.217683  0.192309  0.442781   
am677_gregory                              0.514275  0.490028  0.329834   
am132_laxdoela_saga                        0.704658  0.655091  0.415481   
am302_landslog                             0.188759  0.183106  0.540222   
am178_thidreks_saga                        0.159823  0.181377  0.140362   
am132_bandamanna_saga                      0.662742  0.624821  0.396665   
am655_laeknisbok                           0.141468  0.130084  0.532872   
am519a_alexanders_saga                     0.242227  0.235070  0.145982   
holmPerg34_landslog                        0.610395  0.621886  0.367845   
am162balpha_njals_saga                     0.574630  0.552985  0.293509   
am113_islendingabok                        0.164039  0.128767  0.435181   
nraNorrFragm7_landslog                     0.218563  0.208716  0.541149   
nks235g_konungs_skuggsja                   0.176983  0.159170  0.650697   
am132_hallfredar_saga                      0.625671  0.593818  0.357697   
am35_heimskringla1                         0.244285  0.202023  0.718933   
am242_codex_wormianus                      0.749087  0.667435  0.429172   
dg8II_olafs_saga                           0.229500  0.207135  0.750518   
am28_codex_runicus                         0.297472  0.266576  0.1159

That's not a good sign, two editions of the _Norwegian Homily Book_ doing no better than `0.54` similarity...